extensions/elasticsearch/elasticsearch-7/src/main/java/com/hazelcast/jet/elastic/ElasticSinkBuilder.java

```diff
@@ -0,0 +1,239 @@
+     * @see <a href="https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-update-settings.html#bulk">
+     * Bulk indexing usage in Elastic documentation</a>
+     */
+    @Nonnull
+    public ElasticSinkBuilder<T> bulkRequestFn(@Nonnull SupplierEx<BulkRequest> bulkRequestFn) {
+        this.bulkRequestFn = checkNonNullAndSerializable(bulkRequestFn, "bulkRequestFn");
+        return this;
+    }
+
+    /**
+     * Set the function mapping the item from a pipeline item to an index request
+     * <p>
+     * For example, to create an IndexRequest for a versioned document:
+     * <pre>{@code
+     * builder.mapToRequestFn((mapItem) ->
+     *                      new IndexRequest("my-index")
+     *                              .source(map)
+     *                              .version((Long) map.get("version"))
+     * }</pre>
+     *
+     * This parameter is required.
+     *
+     * @param mapToRequestFn maps an item from the stream to an {@link org.elasticsearch.action.index.IndexRequest},
+     *                       {@link org.elasticsearch.action.update.UpdateRequest} or
+     *                       {@link org.elasticsearch.action.delete.DeleteRequest}
+     * @param <T_NEW> type of the items from the pipeline
+     */
+    @Nonnull
+    @SuppressWarnings("unchecked")
+    public <T_NEW> ElasticSinkBuilder<T_NEW> mapToRequestFn(
+            @Nonnull FunctionEx<? super T_NEW, ? extends DocWriteRequest<?>> mapToRequestFn
+    ) {
+        ElasticSinkBuilder<T_NEW> newThis = (ElasticSinkBuilder<T_NEW>) this;
+        newThis.mapToRequestFn = checkNonNullAndSerializable(mapToRequestFn, "mapToRequestFn");
+        return newThis;
+    }
+
+    /**
+     * Set the function that provides {@link RequestOptions}
+     * <p>
+     * It can either return a constant value or a value based on provided request.
+     * <p>
+     * For example, to provide a custom authentication header:
+     * <pre>{@code
+     * sinkBuilder.optionsFn((request) -> {
+     *     RequestOptions.Builder builder = RequestOptions.DEFAULT.toBuilder();
+     *     builder.addHeader("Authorization", "Bearer " + TOKEN);
+     *     return builder.build();
+     * })
+     * }</pre>
+     *
+     * @param optionsFn function that provides {@link RequestOptions}
+     * @see <a
+     * href="https://www.elastic.co/guide/en/elasticsearch/client/java-rest/current/java-rest-low-usage-requests.html">
+     * RequestOptions in Elastic documentation</a>
+     */
+    @Nonnull
+    public ElasticSinkBuilder<T> optionsFn(@Nonnull FunctionEx<? super ActionRequest, RequestOptions> optionsFn) {
+        this.optionsFn = checkNonNullAndSerializable(optionsFn, "optionsFn");
+        return this;
+    }
+
+    /**
+     * Set the local parallelism of this sink. See {@link SinkBuilder#preferredLocalParallelism(int)}.
+     */
+    @Nonnull
+    public ElasticSinkBuilder<T> preferredLocalParallelism(int preferredLocalParallelism) {
```

Comment:
This should not be an option for the sink, it's not offered for any other sink. We just set a reasonable default.